### PR TITLE
Finish query only after the results are fully consumed

### PR DESCRIPTION
**Pull request by arhimondr**
This commit makes it consistent with QueryState#FINISHED documentation.

This change is needed to allow spooling exchange to be used for query
results delivery. Spooling exchanges in fault tolerant execution are
closed on query completion. It is important not to close an exchange
before the results are fully consumed.

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine

> How would you describe this change to a non-technical end user or system administrator?

With this change a query is considered finished successfully only when the query results are successfully consumed by the client. Before it was possible that a query might've been marked as successful even when client experienced a failure or dropped a connection.

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Core
* Queries are now marked as finished successfully only when the query results are successfully consumed by the client
```

### CODE REVIEW

**Comment by arhimondr** [comment]
### Why this change is needed?

In fault tolerant execution we've been using a streaming (direct) exchange to deliver query results to coordinator. With https://github.com/trinodb/trino/pull/13025 query results will be delivered using a spooling exchange. Spooling exchanges are closed (wiped out) upon query completion (either success or a failure). It is important not to close a spooling exchange before the query results are fully consumed. To achieve that it is required to only transition a query to a `FINISHED` state when the results are fully consumed.

This change also makes the behavior consistent with the `QueryState#FINISHED` documentation: https://github.com/trinodb/trino/blob/6cb319916328184a673d3ca9eb204cc5181c9f25/core/trino-main/src/main/java/io/trino/execution/QueryState.java#L52

**Comment by losipiuk** [comment]
What would happen if a client on purpose reads only a fraction of the result. I think this may be a common use case that a UI tool interacting with Trino (over JDBC or ODBC) just reads a fixed number of rows, up to some limit, even if the result is larger. Would query be marked as FAILED then in Trino? It should not be.

File: core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java [hunk 1]
```diff
@@ -165,6 +165,9 @@
     private Supplier<DynamicFiltersStats> dynamicFiltersStatsSupplier = () -> DynamicFiltersStats.EMPTY;
     private final Object dynamicFiltersStatsSupplierLock = new Object();
 
+    private final AtomicBoolean committed = new AtomicBoolean();
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:168]: `committed` is confusing - you think transactions - while this is not related here at all.
Maybe lets have more straightforward names here:
 * resultsReady
 * resultsConsumed
?

File: core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java [hunk 2]
```diff
@@ -937,13 +941,28 @@ public void onFailure(Throwable throwable)
...
+            committed.set(true);
             transitionToFinished();
         }
         return true;
     }
 
+    public void resultsConsumed()
+    {
+        consumed.set(true);
+        transitionToFinished();
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:953]: transtionToFailed now has conditional logic inside and may or may not do actual transition.
Maybe renamt to `transtionToFinishedIfReady`?

File: core/trino-main/src/main/java/io/trino/server/protocol/Query.java [hunk 3]
```diff
@@ -181,7 +185,7 @@ public static Query create(
         result.queryManager.addOutputInfoListener(result.getQueryId(), result::setQueryOutputInfo);
 
         result.queryManager.addStateChangeListener(result.getQueryId(), state -> {
-            if (state.isDone()) {
+            if (state.isDone() || state == FINISHING) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/protocol/Query.java:188]: looks like this should be separate commit.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/server/protocol/Query.java:188]: Or I guess I do not fully understand this change. Can you explain?

**Review by losipiuk — COMMENTED** [review]
LGTM but let's sync so I am sure I understand it correctly.

File: core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java (hunk 1 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:168]: > committed is confusing - you think transactions - while this is not related here at all.

It is actually about transactions. Basically before considering a query as `FINISHED` it has to finish active execution (be in the `FINISHING` state), commit a transaction if one is active and auto commit is enabled (`commited`) and in addition to that with this PR it would be required for all the results to be consumed (`resultsConsumed`).

File: core/trino-main/src/main/java/io/trino/server/protocol/Query.java (hunk 3 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/server/protocol/Query.java:188]: That's a tricky one. For queries with no output stage (e.g.: DDL operations such as `DROP TABLE`) the exchange is still created. It has to be closed to allow query to finish, otherwise the client will be waiting of `ExchangeClient#isBlocked` and the query will never finish. The client is closed by the `closeExchangeClientIfNecessary`. However with this change the query will not transition to the `FINISHED` state before the results are consumed. So we have to close the `ExchangeClient` when a query is in `FINISHING` state to allow the client to pool the results and mark the query as `resultsConsumed` to allow it to finish.

File: core/trino-main/src/main/java/io/trino/server/protocol/Query.java [hunk 4]
```diff
@@ -181,7 +181,7 @@ public static Query create(
         result.queryManager.addOutputInfoListener(result.getQueryId(), result::setQueryOutputInfo);
 
         result.queryManager.addStateChangeListener(result.getQueryId(), state -> {
-            if (state.isDone()) {
+            if (!result.exchangeClient.isFinished()) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/protocol/Query.java:184]: Actually I prefered previous shape with explanation you put in the PR; maybe just convert it to comment in the code.
No we are closing exchange on ANY state change - which is weird. The question which arises why do we need to open it at all in the first place.

**Review by losipiuk — APPROVED** [review]
